app/3.0/toolbox/service-worker.md

```diff
@@ -2,104 +2,144 @@
-Worker](https://developers.google.com/web/fundamentals/primers/service-worker/) on Web Fundamentals.
+Workers](https://developers.google.com/web/fundamentals/primers/service-worker/) on Web Fundamentals.
 
 ## Prerequisites
 
-To work with service worker, your application **must** be served over HTTPS. However, you can
-test service worker on your local system without a SSL certificate, because `localhost` is
+To work with a service worker, your application **must** be served over HTTPS. However, you can
+test your service worker locally without a SSL certificate, because `localhost` is
 considered a secure origin.
 
-## Add a service worker
-
-Support for Service Worker Precache is built into the [Polymer CLI](/{{{polymer_version_dir}}}/docs/tools/polymer-cli),
-so a service worker script is automatically generated when you build your app.
-
-However, to use the service worker, you need to add code to register it:
-
-```js
-// Register service worker if supported.
-if ('serviceWorker' in navigator) {
-  navigator.serviceWorker.register('/service-worker.js');
+For information on browser support for service worker, see [Is Service Worker Ready](https://jakearchibald.github.io/isserviceworkerready/).
+
+## Add a service worker to your build
+
+To add a service worker to your build:
+
+1.  [Configure polymer.json](#configpolymerjson).
+2.  [Add code to your entrypoint to register a service worker](#register).
+3.  Run `polymer build`.
+
+A service worker is added to your build.
+
+### Configure polymer.json
+
+To configure polymer.json, set the `"entrypoint"` and `"shell"` properties. The service worker will automatically precache these resources.
+
+Polymer CLI generates a service worker by default. You can switch service worker generation on or off with the `"addServiceWorker"` option in your build configuration.
+
+Example polymer.json configuration {.caption}
+
+```json
+{
+  "entrypoint": "index.html",
+  "shell": "src/my-app.js",
+  "builds": [
+    {
+      "name": "esm-unbundled",
+      "browserCapabilities": [
+        "es2015",
+        "modules"
+      ],
+      "js": {
+        "minify": true
+      },
+      "css": {
+        "minify": true
+      },
+      "html": {
+        "minify": true
+      },
+      "bundle": false,
+      "addServiceWorker": true
+    }
+  ],
+  ...
 }
+``` 
+
+### Register your service worker
+
+To use the service worker, add code to your app's entrypoint to register it:
+
+Register a service worker {.caption}
+
+```html
+<head>
+  ...
+  <script>
+  // Feature detect for service worker capability in the browser
+  if ('serviceWorker' in navigator) {
+    // Delay registering until page load
+    window.addEventListener('load', function() {
+      navigator.serviceWorker.register('service-worker.js', {
+        // See rootPath docs https://www.polymer-project.org/3.0/docs/devguide/settings
+        scope: Polymer.rootPath
```

Comment:
Opps, thanks. Does the psk source also need to change?

https://github.com/Polymer/polymer-starter-kit/blob/0f145c34ae4272d60d5c2f7cd93b3acdd61c83fd/index.html#L81

Reply:
my bad. Just before this line, the psk does this: `window.Polymer = { rootPath: '/' };`

Reply:
It will, yeah. But that code creates a `Polymer` object just above the call to `register`:

https://github.com/Polymer/polymer-starter-kit/blob/0f145c34ae4272d60d5c2f7cd93b3acdd61c83fd/index.html#L75

In 2.x, Polymer can read this object and use the value to set its `rootPath` property... But that doesn't work in Polymer 3.x, because modules.

Reply:
Filed a bug for PSK: https://github.com/Polymer/polymer-starter-kit/issues/1110

Reply:
Thanks! That makes more sense